File: src/image_stitching.cpp
#include "opencv_all.hpp"

int main(void)
{
    // Load two images (c.f. Assume that two images have the same size and type)
    cv::Mat image1 = cv::imread("data/hill01.jpg");
    cv::Mat image2 = cv::imread("data/hill02.jpg");
    if (image1.empty() || image2.empty()) return -1;

    // Retrieve matching points
    cv::Mat gray1, gray2;
    if (image1.channels() > 1)
    {
        cv::cvtColor(image1, gray1, CV_RGB2GRAY);
        cv::cvtColor(image2, gray2, CV_RGB2GRAY);
    }
    else
    {
        gray1 = image1.clone();
        gray2 = image2.clone();
    }
    cv::Ptr<cv::FeatureDetector> detector = cv::xfeatures2d::SURF::create(); // SURF features
    std::vector<cv::KeyPoint> keypoint1, keypoint2;
    detector->detect(gray1, keypoint1);
    detector->detect(gray2, keypoint2);
    cv::Ptr<cv::FeatureDetector> extractor = cv::xfeatures2d::SURF::create(); // SURF descriptors
    cv::Mat descriptor1, descriptor2;
    extractor->compute(gray1, keypoint1, descriptor1);
    extractor->compute(gray2, keypoint2, descriptor2);
    cv::FlannBasedMatcher matcher; // Approximate Nearest Neighbors (ANN) matcher
    std::vector<cv::DMatch> match;
    matcher.match(descriptor1, descriptor2, match);

    // Calculate planar homography and merge them
    std::vector<cv::Point2f> points1, points2;
    for (size_t i = 0; i < match.size(); i++)
    {
        points1.push_back(keypoint1.at(match.at(i).queryIdx).pt);
        points2.push_back(keypoint2.at(match.at(i).trainIdx).pt);
    }
    cv::Mat H = cv::findHomography(points2, points1, cv::RANSAC);
    cv::Mat merge;
    cv::warpPerspective(image2, merge, H, cv::Size(image1.cols * 2, image1.rows));
    merge.colRange(0, image1.cols) = image1 * 1; // Copy

    // Show the merged image
    cv::imshow("3DV Tutorial: Image Stitching", merge);
    printf("Press any key to terminate tihs program!\n");
    cv::waitKey(0);
    return 0;
}

File: scripts/image_stitching.py
import numpy as np
import cv2

def main():
    # Load two images
    input1, input2 = "../bin/data/hill01.jpg", "../bin/data/hill02.jpg"
    image1 = cv2.imread(input1)
    image2 = cv2.imread(input2)

    # Retrieve matching points
    brisk = cv2.BRISK_create()
    keypoints1, descriptors1 = brisk.detectAndCompute(image1, None)
    keypoints2, descriptors2 = brisk.detectAndCompute(image2, None)

    fmatcher = cv2.DescriptorMatcher_create("BruteForce-Hamming")
    match = fmatcher.match(descriptors1, descriptors2)

    # Calculate planar homography and merge them
    points1, points2 = [], []
    for i in range(len(match)):
        points1.append(keypoints1[match[i].queryIdx].pt)
        points2.append(keypoints2[match[i].trainIdx].pt)
    
    points1 = np.array(points1, dtype=np.float32)
    points2 = np.array(points2, dtype=np.float32)

    H, inlier_mask = cv2.findHomography(points2, points1, cv2.RANSAC)
    merged = cv2.warpPerspective(image2, H, (image1.shape[1]*2, image1.shape[0]))
    merged[:,:image1.shape[1]] = image1
    merged[:,0:image1.shape[1]] = image1 # copy
    cv2.imshow("3DV Tutorial: Image Stitching", merged)

    # show the merged image
    matched = cv2.drawMatches(img1=image1, 
                                keypoints1=keypoints1, 
                                img2=image2, 
                                keypoints2=keypoints2, 
                                matches1to2=match[:15], 
                                outImg=None)
    
    original = np.hstack((image1, image2))
    matched = np.vstack((original, matched))
    merged = np.vstack((matched, merged))
    
    cv2.imshow("3DV Tutorial: Image Stitching", merged)
    cv2.waitKey(0)

if __name__ == "__main__":
    main()